Making memref to capture vectorC/D in gemm from xdlops lowering

jerryyin opened the pull request:
First commit:
 - Amended blockwisegemm and xdlopsgemm op signature to include return
 - Fixed xdlopsgemm out of bound access when kPack = 1 and k_base > 1
 - Moved xdlops register offset to attributes
 - Moved gridwisegemm allocation before blockwisegemm compuation

Second commit:
 - Replaced vectorC/D as memref in blockwise/xdlops gemm

Third commit: 
 - Added vectorC/D type conversion before write back

Forth commit
 - Amended memref of scalars to memref of vectors
 - Fixed the redundantly allocated vgpr in populated ISA

Fifth commit
 - Fixed all unit tests failures

jerryyin committed:
Adding memref argument to capture gemm returns

- Amended blockwisegemm and xdlopsgemm op signature to include return
- Fixed xdlopsgemm out of bound access when kPack = 1 and k_base > 1
- Moved xdlops register offset to attributes
- Moved gridwisegemm allocation before blockwisegemm compuation

jerryyin committed:
Replaced vectorC/D as memref in blockwise/xdlops gemm

jerryyin committed:
Adding register C type conversion

- Fixed MIOpenOps.td for xdlopsgemm as changes get lost over rebase
- Added register C type conversion to fix fp16 and bf16 sanity tests

jerryyin committed:
Converted memref of scalars to vectors

jerryyin committed:
fixing unit test failures

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -736,7 +737,7 @@ def MIOpen_BlockwiseGemmV2Op:
   }];
   let assemblyFormat = [{
-    `(` operands `)` attr-dict `:` type(operands) `->` type($vectorDs)
+    `(` operands `)` attr-dict `:` type(operands)
```
Could I talk you into creating a pretty assembly format here?
Something like
```
$matrixC `+` `=` $matrixA `*` $matrixB
```
?
It's not _required_, but since we're here...

jerryyin replied:
Good idea! Indeed I should, and I will actually amend (or make sure) all gemm signatures are in this consistent format.

krzysz00 replied:
I've been amending them as I refactor ... which reminds me, I should make sure I got `gridwise_gemm`

krzysz00 replied:
One note, if you're writing `+=`, you want 
```
`+` `` `=`
```
to get rid of a random space

jerryyin replied:
Sounds good, I was following your original practice anyways.

So if I do `+=` it automatically add a space in between? Haha, wired.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -769,8 +770,8 @@ def MIOpen_XdlopsGemmV2Op:
-                   Index:$regOffsetA,
-                   Index:$regOffsetB,
+                   IndexAttr:$regOffsetA,
+                   IndexAttr:$regOffsetB,
```
Thanks!

jerryyin replied:
No problem. If I don't do this, we'd have to generate redundant IRs to calculate the right offset in this PR. Now looking forward to get rid of those in upcoming PRs...

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -789,7 +790,7 @@ def MIOpen_XdlopsGemmV2Op:
   }];
   let assemblyFormat = [{
-    `(` operands `)` attr-dict `:` type(operands) `->` type(results)
+    `(` operands `)` attr-dict `:` type(operands)
```
Simalarly on assembly format

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -2331,9 +2336,8 @@ struct GridwiseGemmV2RewritePattern
-    int64_t resultCVectorLen = vectorType.getNumElements();
-    int64_t numElements = resultCVectorLen * tailResults.size();
 
+    int64_t numElements = regCVectorLen * vectorNumber;
```
Can we name `vectorNumber` something like `numVectors`?

jerryyin replied:
Yes agreed that this is an awkward name, but I was following the naming convention from `XdlopsSelection.h`. To limit the scale of refactoring in this PR, I'd rename the local variable name and leave the more comprehensive renaming to when we touch `XdlopsSelection.h`

krzysz00 replied:
Agreed

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -2512,32 +2516,50 @@ struct GridwiseGemmV2RewritePattern
+
+    // Convert from memref<?xvector<?xT>> to memref<?xT> where the source T
+    // is the accumulatorType and destination type is destType
+    auto convertLoop = b.create<TransformingForOp>(
```
Instead of the `MulIOp`, can we have the `bounds` be the size of the destination memref and the `strides` be the size of the vector?

jerryyin replied:
Nice suggestion! I don't like this `MulIOp` to start with, I'm glad you point this out to me, thank you!

jerryyin replied:
Wait, how does it work to make the source and destination in two coordinate space?

This transforming for op would:
 - `load registerC[coord]`
 - `type convert`
 - `store loaded convertedC[coord * vectorLen]` 

What should I do such that the load on registerC is still on linear coordinate space whereas store is strided by `vectorLen`

krzysz00 replied:
Good question. Earlier idea withdrawn. My gut is saying to make a quick map that `Embed`s the destination coordinate with the appropriate coefficient to scale it from `[0, 1, ...]` to `[0, L, 2L, ...]` so that the multiplication is part of the loop structure, and thus analyzable.

krzysz00 replied:
To rephrase, instead of having one iteration domain, have two
The second one is
```
int64_t numOutputs = vectorType.getNumElements() * numVectors;
BottomUpTMBuilder vectorToScalar(b, {"scalar"}, {numOutputs}, loc);
vectorDest.embed({"vector"}, {0}, {numVectors}, "vector", {vectorType.getNumElements()});
TransformMapAttr vectorToScalarAttr = vectorToScalar.get();

...
// Set up the TransformingFor to take both the no-transforms array (vector coordinates) and the array containing the above transform (which gives scalar coordinates)
```

jerryyin replied:
Addressed in c59cfa8

krzysz00 reviewed (commented):
I've got a few initial comments - the main one is the loop you've added to gridwise gemm

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -2222,15 +2221,31 @@ struct GridwiseGemmV2RewritePattern
+    }
+
+    VectorType accumulatorVectorType =
+        vectorType.cloneWith({}, accumulatorType);
```
Is this spelled `clone(accumulatorType)` also?

jerryyin replied:
Hmm you are right, it looks like clone is the high level wrapper that allows either shape or type argument.

jerryyin committed:
Addressing review feedbacks

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -736,7 +739,10 @@ def MIOpen_BlockwiseGemmV2Op:
   }];
   let assemblyFormat = [{
-    `(` operands `)` attr-dict `:` type(operands) `->` type($vectorDs)
+    $matrixC `+` `` `=` $bufferA `by` $matrixA `[` $waveOffsetA `]` `*`
```
Nit: is `by` the right word here? Maybe `from` or `<-`

jerryyin replied:
Yes, `from` is better, I used `by` just so that it is as short as possible.

krzysz00 replied:
I'd still go for `from` or something like `<-` on the theory that it'll make reading the IR clearer, which is rather important.

jerryyin replied:
Addressed in c59cfa8

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseGemmLowering.cpp:
```diff
@@ -198,12 +201,12 @@ struct XdlopsGemmV2RewritePattern : public OpConversionPattern<XdlopsGemmV2Op> {
-    auto bufferBType = adaptor.bufferB().getType().cast<MemRefType>();
-    Type bufferAElementType = bufferAType.getElementType();
-    Type bufferBElementType = bufferBType.getElementType();
+    Value matrixA = adaptor.matrixA();
```
I see things seem to have been renamed, but the one place I didn't see a renaming is in the names of the operation arguments themselves?

jerryyin replied:
If I understand your question right, there used to be a `matrixA` argument, and whose name I reused (and replaced `bufferA`). So it is for `matrixB` and `bufferB`. You can see the updated diff on `MIOpenOps.td` that has those renames.

krzysz00 replied:
`MIOpenOps.td` doesn't rename the arguments at all.

What I was trying to work out: was it the case that, before this, "buffer" meant the LDS memref and "matrix" meant the in-register stuff and now you've switched it? It feels like you've switched it

jerryyin replied:
See [here](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/711/files#r940651714) for the rename.

Previously, matrix was lds and buffer is register. Now, matrix is register (and lds is no longer needed in new xdlops because blockwisegemm did the copy already).

krzysz00 reviewed (commented):
Ok, so, we're definitely in the trivial nitpicking zone here - the main bit I'm staring at is the changes between what's called "buffer" and what's called "matrix"

jerryyin commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -765,22 +771,18 @@ def MIOpen_ThreadwiseGemmOp:
-                   Index:$regOffsetB,
+    Arguments<(ins IndexAttr:$regOffsetA,
+                   IndexAttr:$regOffsetB,
                    MemRefOf<[F32, F16, BF16, I8,
                              VectorOfLengthAndType<[2, 4], [F32]>,
                              VectorOfLengthAndType<[2, 4, 8], [F16, BF16]>,
-                             VectorOfLengthAndType<[4, 8, 16], [I8]>]>:$bufferA,
+                             VectorOfLengthAndType<[4, 8, 16], [I8]>]>:$matrixA,
                    MemRefOf<[F32, F16, BF16, I8,
                              VectorOfLengthAndType<[2, 4], [F32]>,
                              VectorOfLengthAndType<[2, 4, 8], [F16, BF16]>,
-                             VectorOfLengthAndType<[4, 8, 16], [I8]>]>:$bufferB,
-                   Variadic<VectorOfRankAndType<[1], [F32, F16, I32]>>:$vectorCs)>,
-    Results<(outs Variadic<VectorOfRankAndType<[1], [F32, F16, I32]>>:$vectorDs)> {
+                             VectorOfLengthAndType<[4, 8, 16], [I8]>]>:$matrixB,
+                   MemRefOf<[F32, F16, BF16, I32,
+                             VectorOf<[F32, F16, BF16, I32]>]>:$matrixC)>{
```
anchor

krzysz00 replied:
... right, I was looking at blockwise gemm's names. I'm very good at this.

krzysz00 reviewed (commented):
Just thought I'd go back and elaborate on some things here

jerryyin committed:
Address review feedbacks

- Amended blockwisegemmv2 tablegen
- Used coordinate transform in vector to scalar conversion

krzysz00 reviewed (approved):
LGTM, thank you!
